src/libs/gstreamer/lv_gstreamer.c

```diff
@@ -0,0 +1,639 @@
+    .constructor_cb = lv_gstreamer_constructor,
+    .destructor_cb = lv_gstreamer_destructor,
+    .width_def = LV_SIZE_CONTENT,
+    .height_def = LV_SIZE_CONTENT,
+    .instance_size = sizeof(lv_gstreamer_t),
+    .base_class = &lv_image_class,
+    .name = "gstreamer",
+};
+
+/**********************
+ *      MACROS
+ **********************/
+
+#if LV_COLOR_DEPTH == 16
+    #define GST_FORMAT   "RGB16"
+    #define IMAGE_FORMAT LV_COLOR_FORMAT_RGB565
+#elif LV_COLOR_DEPTH == 24
+    #define GST_FORMAT   "BGR"
+    #define IMAGE_FORMAT LV_COLOR_FORMAT_RGB888
+#elif LV_COLOR_DEPTH == 32
+    #define GST_FORMAT   "BGRA"
+    #define IMAGE_FORMAT LV_COLOR_FORMAT_ARGB8888
+#else
+    #error Unsupported LV_COLOR_DEPTH
+#endif
+
+/**********************
+ *   GLOBAL FUNCTIONS
+ **********************/
+
+lv_obj_t * lv_gstreamer_create(lv_obj_t * parent)
+{
+    if(!gst_is_initialized()) {
+        gst_init(0, NULL);
+    }
+
+    LV_TRACE_OBJ_CREATE("begin");
+
+    lv_obj_t * obj = lv_obj_class_create_obj(MY_CLASS, parent);
+    lv_obj_class_init_obj(obj);
+
+    LV_TRACE_OBJ_CREATE("end");
+    return obj;
+}
+
+lv_result_t lv_gstreamer_set_src(lv_obj_t * obj, const char * factory_name, const char * property, const char * source)
+{
+    LV_ASSERT_NULL(obj);
+    LV_ASSERT_NULL(factory_name);
+    LV_ASSERT_OBJ(obj, MY_CLASS);
+
+    if(!obj || !factory_name) {
+        LV_LOG_WARN("Refusing to set source with invalid params. Obj: %p Factory Name: %s", obj, factory_name);
+        return LV_RESULT_INVALID;
+    }
+
+    lv_gstreamer_t * streamer = (lv_gstreamer_t *)obj;
+
+    if(streamer->pipeline) {
+        LV_LOG_WARN("LVGL doesn't allow modifying the GStreamer source. Create a new widget with a new src instead");
+        return LV_RESULT_INVALID;
+    }
+    GstElement * pipeline = gst_pipeline_new("lv_gstreamer_pipeline");
+    if(!pipeline) {
+        LV_LOG_ERROR("Failed to create gstreamer pipeline");
+        return LV_RESULT_INVALID;
+    }
+
+    GstElement * head = gst_element_factory_make(factory_name, "lv_gstreamer_source");
+    if(!head) {
+        gst_object_unref(pipeline);
+        LV_LOG_ERROR("Failed to create source from factory '%s'", factory_name);
+        return LV_RESULT_INVALID;
+    }
+
+    if(!gst_bin_add(GST_BIN(pipeline), head)) {
+        gst_object_unref(pipeline);
+        LV_LOG_ERROR("Failed to add source element to pipeline");
+        return LV_RESULT_INVALID;
+    }
+
+    if(property != NULL && source != NULL) {
+        g_object_set(G_OBJECT(head), property, source, NULL);
+    }
+
+    /* The uri decode source element will automatically handle parsing and decoding for us
+     * for other source types, we need to add a parser and a decoder ourselves element*/
+    if(!lv_streq(LV_GSTREAMER_FACTORY_URI_DECODE, factory_name)) {
+        GstElement * decodebin = gst_element_factory_make("decodebin", "lv_gstreamer_decodebin");
+        if(!decodebin) {
+            gst_object_unref(streamer->pipeline);
```

Comment:
`streamer->pipeline` isn't assigned until the end of this function. Update the other 2 cases of this too.

```suggestion
            gst_object_unref(pipeline);
```

Reply:
Argh, nice catch. I originally went with assigning everything to `streamer->pipeline` and modified it along the way but didn't change all of these unref calls